A83-xds-gcp-authn-filter.md

```diff
@@ -0,0 +1,218 @@
+a call attribute for access in filters.  We will now also pass the CDS
+resource for the cluster in a separate call attribute, so that the GCP
+Authentication filter can access the cluster metadata for the selected
+cluster.  Note that this will require implementing [A74], so that the
+CDS resource is available at xDS routing time.
+
+### Filter Behavior
+
+The filter will maintain a cache of
+GcpServiceAccountIdentityCallCredentials instances, one for each
+audience, along with a last-used timestamp for each one.  The maximum
+number of entries in the cache is bounded by the config field
+`cache_config.cache_size`; if the cache exceeds that size, then the
+entry with the oldest last-used timestamp will be removed.
+
+When the filter processes the RPC's initial metadata, it will first
+attempt to determine the audience for the request by looking at the
+cluster metadata key corresponding to the filter instance name, which
+must be of type `extensions.filters.http.gcp_authn.v3.Audience`.  If the
+cluster has no such metadata key, the filter is a no-op.  Otherwise, the
+audience is the value of the `url` field in the `Audience` proto.
+
+The filter will then check to see if it already has a cached
+GcpServiceAccountIdentityCallCredentials instance for the specified
+audience.  If it does not, it will create a new instance, adding it to
+its cache, removing the least recently used entry from the cache if the
+cache is already at its max size.  It will then attach that
+GcpServiceAccountIdentityCallCredentials instance to the RPC.
+
+Note that implementations must ensure that the token is not added to
+RPCs sent on insecure connections.  However, the GCP Authentication
+filter will run before load balancing has chosen a connection, so the
+filter cannot directly add the token to the RPC.  Instead, it must add
+the call credential to the RPC, and the call credential will do the work
+of adding the token to the RPC later, after load balancing has chosen a
+connection.
+
+### Temporary environment variable protection
+
+Support for the GCP Authentication filter in LDS will be guarded by the
+`GRPC_EXPERIMENTAL_XDS_GCP_AUTHENTICATION_FILTER` env var.  The env var
+guard will be removed once the feature passes interop tests.
+
+## Rationale
+
+It is not our intention to support this mechanism for GCP only; in
+principle, it should be possible to support JWT identity tokens for any
+cloud provider.  However, at present, the existing xDS HTTP filter
+supports only GCP, so that's what we're initially focusing on, for
+compatibility with Envoy.  We would be open to future contributions from
+the OSS communicate to provide similar functionality for other cloud
+providers, in both gRPC and Envoy.
+
+Note that having the filter cache
+GcpServiceAccountIdentityCallCredentials instances instead of directly
+caching JWT tokens means that the cache semantics will be slightly
+different than Envoy: cache expiration will be based solely on
```

Comment:
"cache" here is imprecise, as there's two caches: one in the filter, one in the cred. And you most recently talked about "caching JWT tokens." But this is also sorta wrong: there are two cache expirations (thus "cache expiration _is not_ based solely on LRU") whereas in Envoy there's only one. I think if you specify which cache, and describe both caches, in order to contrast it.

Reply:
Good point.  Rewrote this to improve clarity.